lib/spack/spack/stage.py

```diff
@@ -381,8 +381,13 @@ def fetch(self, mirror_only=False):
             # Add URL strategies for all the mirrors with the digest
             for url in urls:
                 fetchers.insert(
-                    0, fs.URLFetchStrategy(
+                    0, fs.from_url_scheme(
                         url, digest, expand=expand, extension=extension))
+                # fetchers.insert(
+                #     0, fs.URLFetchStrategy(
+                #         url, digest, expand=expand, extension=extension))
```

Comment:
Here, the URL `scheme`-based function is used to determine what fetch strategy to employ.

Question: In testing, I noticed that if one the strategies throws, the whole fetch operation fails.  I understand that what is preferred is that every URL should be tried (with their associated fetch strategies), and only if they *all* fail to download the source should the fetch operation fail.  What do others think?  If this behavior is, indeed, preferred, where would be the best places to make the necessary changes?

Reply:
I think this change, as-is, is the likely cause of #13404.

My intent was, put simply, to allow fetching from `s3://` URLs by providing `S3FetchStrategy`.  The issue is that the original logic here, where the fetch strategy is selected, was problematic because it effectively always used `URLFetchStrategy` before trying anything else, and appending the S3 strategy did not work because, as I noted above, the whole fetch operation would fail when the first attempt using `URLFetchStrategy` raised an exception.

What I ended up doing was just checking the URLs `scheme` and matching it against a known table of fetch strategies.  This change may have led to the issue found in #13404, though I'm not yet sure how.

Reply:
I thought a little more about the changes, and I think fetching web URLs are now going through urllib (with these changes).  I believe before these changes spack was running `curl` to perform the downloads.  I'm not 100% sure that my recollection is correct, but if it is, that difference is very likely to be the specific change that introduced #13404.

@scheibelp